File: kubik/kubik-server/src/main/java/com/cspinformatique/kubik/server/domain/warehouse/service/StocktakingProductService.java
package com.cspinformatique.kubik.server.domain.warehouse.service;

import java.util.List;

import com.cspinformatique.kubik.server.model.product.Product;
import com.cspinformatique.kubik.server.model.warehouse.StocktakingProduct;
import com.cspinformatique.kubik.server.model.warehouse.Stocktaking.Status;

public interface StocktakingProductService {
	StocktakingProduct addProductToCategory(int productId, long categoryId);
	
	int countCategoriesWithProduct(int productId, long categoryId);
	
	void delete(long id);

	List<StocktakingProduct> findByProductAndStocktakingStatus(Product product, Status status);
	
	StocktakingProduct updateQuantity(long id, double quantity);
	
	StocktakingProduct save(StocktakingProduct stocktakingProduct);
}

File: kubik/kubik-server/src/main/webapp/public/js/warehouse/stocktaking-product-service.js
(function() {
	var STOCKTAKING_PRODUCT_URL = "/stocktaking-product"
	var STOCKTAKING_URL = "/stocktaking";

	angular
		.module("Kubik")
		.factory("stocktakingProductService", StocktakingProductService);

	function StocktakingProductService($http) {
		return {
			addProductToCategory : addProductToCategory,
			deleteStocktakingProduct : deleteStocktakingProduct,
			updateQuantity : updateQuantity
		};
		
		function addProductToCategory(productId, stocktakingCategoryId, stocktakingId){
			return $http
				.post(
						STOCKTAKING_URL + "/" + stocktakingId + "/stocktakingCategory/" + 
						stocktakingCategoryId + "/stocktakingProduct?productId=" + productId)
				.then(postSuccess);
			
			function postSuccess(response){
				return response.data;
			}
		}
		
		function deleteStocktakingProduct(stocktakingProduct){
			return $http
				.delete(STOCKTAKING_PRODUCT_URL + "/" + stocktakingProduct.id);
		}
		
		function updateQuantity(stocktakingProductId, quantity){
			return $http
				.post(STOCKTAKING_PRODUCT_URL + "/" + stocktakingProductId + "?quantity=" + quantity)
				.then(postSuccess);
			
			function postSuccess(response){
				return response.data;
			}
		}
	}
})();